File: 115-distinct-subsequences/115-distinct-subsequences.cpp
class Solution {
public:
    int numDistinct(string s, string t) {
        int m=t.length();
        int n=s.length();
        if(m>n)
        {
            return 0;
        }
        double dp[1001][1001];
        
        for(int i=0;i<=m;i++)
        {
            dp[i][0]=0;
        }
         for(int j=0;j<=n;j++)
        {
            dp[0][j]=1;
        }
        for(int i=1;i<=m;i++)
        {
            for(int j=1;j<=n;j++)
            {
                if(t[i-1]!=s[j-1])
                {
                    dp[i][j]=dp[i][j-1];
                   
                }
                else
                {
                    dp[i][j]=dp[i-1][j-1]+dp[i][j-1]; 
                }
            }
        }
        return dp[m][n];
    }
};

File: 115_Distinct_Subsequences.py
class Solution:
    def numDistinct(self, s, t):
        dp = [[0] * (len(s) + 1) for _ in range(len(t) + 1)]
        for col in range(len(dp[0])):
            dp[0][col] = 1
        for x in range(1, len(s) + 1):
            for y in range(1, len(t) + 1):
                if s[x - 1] == t[y - 1]:
                    dp[y][x] = dp[y - 1][x - 1] + dp[y][x - 1]
                else:
                    dp[y][x] = dp[y][x - 1]
        return dp[-1][-1]
